### PR TITLE
0707/charge and aotu refund; tools

### PR DESCRIPTION
**Pull request by Tigo9527**
This change is [Reviewable](https://reviewable.io/reviews/conflux-chain/web3pay-contract/8)

### CODE REVIEW

File: contracts/APPCoin.sol [hunk 1]
```diff
@@ -59,7 +60,21 @@ contract APPCoin is Initializable, ERC777Upgradeable, PausableUpgradeable, Ownab
             delete frozenMap[acc];
         }
     }
+    function charge(address account, uint256 amount, bytes memory data) public onlyOwner {
```
Comment (wanliqun) [inline contracts/APPCoin.sol:63]: `onlyAppOwner`?

File: contracts/APPCoin.sol [hunk 2]
```diff
@@ -59,7 +60,21 @@ contract APPCoin is Initializable, ERC777Upgradeable, PausableUpgradeable, Ownab
...
+        _burn(account, amount, "", data);
+        if (frozenMap[account] > 1) {
+            // refund
+            uint256 appCoinLeft = balanceOf(account);
+            _burn(account, appCoinLeft, "", "refund");
+            IERC777Upgradeable(apiCoin).send(account, appCoinLeft, "refund");
+            delete frozenMap[account];
+            emit Withdraw(account, appCoinLeft);
+        }
```
Comment (wanliqun) [inline contracts/APPCoin.sol:72]: Does this make any loophole to bypass the time lock?

**Review by wanliqun — APPROVED** [review]

File: contracts/APPCoin.sol (hunk 1 above)
Comment (Tigo9527) [reply contracts/APPCoin.sol:63]: It depends on the implementation of billing service. May change later.

File: contracts/APPCoin.sol (hunk 2 above)
Comment (Tigo9527) [reply contracts/APPCoin.sol:72]: Refunding after charging is normal logic, as discussed.

**Pull request merged**